### DISCUSSION TITLE
How to validate a field directly in v2

### DISCUSSION THREAD

**Discussion by sphuber** [pydantic/pydantic#7367] [Question] [answered]
I have been googling for hours, but have not been able to come up with an answer to the following question.

In pydantic v1, it was possible to do the following:
```python
from pydantic import BaseModel, IntegerError

class Model(BaseModel):
    a: int

field = Model.__fields__['a']
value, error = field.validate('invalid', {}, loc=None)
assert isinstance(error.exc, IntegerError)
value, error = field.validate(1, {}, loc=None)
assert error is None
```
In this minimal example, it seems of course pointless to validate the single field individually directly on the `Field` instance. However, in my actual use-case, the field is part of a heavily nested model, and in some part of the application I need to validate a value for a single (variable) particular field.

How can I reproduce this in pydantic v2? The `BaseModel.fields` mapping now returns `FieldInfo` instances instead of `Field` instances, which don't seem to have any validation method.

Thanks for any suggestions in advance :+1:

**Comment by sydney-runkle** [comment]
Hi @sphuber,

Thanks for your question! One way to approach solving this is using Pydantic's V2 `TypeAdapter` class.  You can read more about them [here](https://docs.pydantic.dev/latest/usage/type_adapter/). In sum, they're super helpful for validating data against objects that aren't `BaseModel`s.

The first few lines of the documentation summarize this well:

> For use cases like this, Pydantic provides TypeAdapter, which can be used for type validation, serialization, and JSON schema generation without creating a BaseModel.
A TypeAdapter instance exposes some of the functionality from BaseModel instance methods for types that do not have such methods (such as dataclasses, primitive types, and more)

---

Here's a solution you could implement for the simple case you listed above:

```py
from pydantic import BaseModel, TypeAdapter, ValidationError


class Model(BaseModel):
    a: int

print(repr(Model.model_fields['a']))
# > FieldInfo(annotation=int, required=True)

# creating a TypeAdapter object based on the type annotation of your designated field
ValidatorA = TypeAdapter(Model.model_fields['a'].annotation)
validated_a = ValidatorA.validate_python(1)
assert validated_a == 1

try:
    invalid_a = ValidatorA.validate_python("invalid")
except ValidationError as exc:
    assert exc.errors()[0]['type'] == "int_parsing"
```

**Reply by jordemort** [reply]
I have the same question. The method you describe only partially works, though:

```python
from pydantic import BaseModel, TypeAdapter, Field

class Model(BaseModel):
    foo: int = Field(gt=7)

foo_validator = TypeAdapter(Model.model_fields['foo'].annotation)

foo_validator.validate_python(5)
print("Hey, that's not greater than 7!")
```

TypeAdapter validates the type, but this method loses the metadata and doesn't validate that.

My use case is a console application that is prompting the user to fill in the fields of a model, one at a time. I would like to be able to validate a value for an individual field of a model, along with all of the constraints that might be imposed by Field, without validating the whole model.

My current plan is to attempt to construct an instance from a dictionary, and then loop through ValidationErrors.errors and ignore anything that doesn't specifically reference the field the user is editing, but it would be much more convenient if I could find a Pydantic 2.x equivalent to `.__fields__[x].validate`.

**Reply by jordemort** [reply, marked as answer]
This feels like a little bit of a hack, but it seems to work like I want it to:

```python
from pydantic import BaseModel, Field

class Model(BaseModel):
    foo: int = Field(gt=7)
    bar: int = Field(lt=7)

Model.__pydantic_validator__.validate_assignment(Model.model_construct(), "foo", 5)
```

It validates the constraint on `foo` without complaining about a missing `bar`:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pydantic_core._pydantic_core.ValidationError: 1 validation error for Model
foo
  Input should be greater than 7 [type=greater_than, input_value=5, input_type=int]
    For further information visit https://errors.pydantic.dev/2.3/v/greater_than
```

**Reply by sphuber** [reply]
Cheers, I will be using this workaround for the time being. Thanks @sydney-runkle as well for your suggestion!

**Reply by rbavery** [reply]
Anyone know if there a better solution nowadays? I want to use my pydantic v2 model to validate individual fields, without creating another non pydantic object for validation. Would like to use something like `BaseModel.model_validate_field()` but that was removed with v2.

**Comment by sarflux** [comment]
Thanks @jordemort , this helped me out a lot for a similar use case, although not entirely. 

For my use case, I have defined a Model that contains numerous fields based on my DB attributes. 

An example model for my use case is as follows:

```python
from pydantic import BaseModel
from typing import Optional
class Model(BaseModel):
    db_id: str
    db_name: str
    db_numeric_attr: Optional[int] = None
```

during create operations, I can validate my data if I have my required attributes, (i.e - "db_id" and "db_name"). However there are instances where I fetch data with multiple attributes but not the required ones, an example would be {"db_id" : "<some ID>", "db_numeric_attr" : 4.0} 
I would like to validate these select fields against my Model without making all my attributes optional. 


so far the answer by @jordemort  helps as I can create a function that takes in my data object and validates each item sequentially, 
example code:

```python
from pydantic import BaseModel
from typing import Dict
def validate_selected_fields_from_model(model:BaseModel, data_dict:Dict):
    for k,v in data_dict.items():
        model.__pydantic_validator__.validate_assignment(model.model_construct(), k, v )
```
however the validation error would occur only once and would exit this function. Would there be a way to validate the entirety of my data with selected fields? The only workaround I can think of is to add error handling, collect all validation errors as they occur and return them.

**Reply by spiritwalking** [reply]
Hi @sarflux , maybe you can use "try-except" to catch exceptions.
In my case, my code is:
```python
from pydantic import BaseModel, ConfigDict, ValidationError, Field
from pydantic import BaseModel
from typing import Dict

class Event(BaseModel):
    model_config = ConfigDict(strict=True)

    when: int
    title: int = Field(default=12, description="Title of this event.")

json_data = {"when": "12.01", "title": "title"}

def validate_selected_fields_from_model(model:BaseModel, data_dict:Dict):
    for k,v in data_dict.items():
        try:
            model.__pydantic_validator__.validate_assignment(model.model_construct(), k, v )
        except ValidationError as e:
            print(e)

validate_selected_fields_from_model(Event, json_data)
```

**Reply by gislerro** [reply]
You can make the function an iterator so you can yield on each assignment. That way an error can propagate as soon as one is encountered

**Reply by lvvittor** [reply]
Just for the record, as @gislerro proposed:

```python
from pydantic import BaseModel, ConfigDict, ValidationError, Field
from typing import Dict, Iterator

class Event(BaseModel):
    model_config = ConfigDict(strict=True)
    
    when: int
    title: int = Field(default=12, description="Title of this event.")

def validate_selected_fields_from_model(model: BaseModel, data_dict: Dict) -> Iterator[ValidationError]:
    """
    Validates fields one by one, yielding any ValidationErrors encountered.
    This allows immediate error handling as soon as any validation fails.
    
    Args:
        model: The Pydantic model to validate against
        data_dict: Dictionary containing data to validate
        
    Yields:
        ValidationError: Any validation errors encountered during the process
    """
    for k, v in data_dict.items():
        try:
            model.__pydantic_validator__.validate_assignment(model.model_construct(), k, v)
        except ValidationError as e:
            yield e
            
 # Example usage
json_data = {"when": "12.01", "title": "title"}

# Process validation errors as they occur
for error in validate_selected_fields_from_model(Event, json_data):
    print(f"Validation error: {error}")
```

**Comment by Kevin-Mc-Callister** [comment]
Would be very cool, what i did for now was:

`    model_config = ConfigDict(validate_assignment=True)
`
Downside is it only takes a dictionary on assignment

**Comment by pycaw** [comment]
There are many who want this properly supported by pydantic, I hope it gets the attention it deserves. My use case is similar to what's been mentioned, the user is to input values per fields and validation/feedback should occur per field as well.

**Comment by lu-pl** [comment]
Hi, I would like to briefly point out that the accepted solution is rather problematic and propose another approach.

The accepted solution by @jordemort essentially creates a model husk using `Model.model_construct` and calls the pydantic-core method `Model.__pydantic_validator__.validate_assignment`. This however runs model validators for every invocation, which are then likely to fail because most fields of the model are not supplied at that point.

Without model validators, the solution works quite well:

```python
from typing import Any
from pydantic import BaseModel, Field

class Point(BaseModel):
    x: int = Field(gt=1, le=3)
    y: int

def validate_model_field(model: type[BaseModel], field: str, value: Any) -> Any:
    result = model.__pydantic_validator__.validate_assignment(
        model.model_construct(), field, value
    )
    return result

validate_model_field(Point, "x", 3)       # valid
validate_model_field(Point, "x", 3.1)     # ValidationError
validate_model_field(Point, "x", 4)       # ValidationError
```

However, with model validators, this is likely to fail with an `AttributeError`, since `__pydantic_validator__` calls model validators with every invocation and due the nature of a *field* validator, not all required model fields are necessarily supplied at the time the validator is called.

```python
from typing import Any
from pydantic import BaseModel, Field, model_validator

class Point(BaseModel):
    @model_validator(mode="after")
    def x_lt_y(self):
        if self.x < self.y:
            raise ValueError("x should be gt y.")
        return self

    x: int = Field(gt=1, le=3)
    y: int

validate_model_field(Point, "x", 2)  # AttributeError: 'Point' object has no attribute 'y'
``` 

The solution I would like to propose utilizes the fact that Pydantic keeps a JSON record of all encountered exceptions during model validation - so the idea is to try to instantiate the model given only the single supplied field and check the exception record for an entry referencing that field; if there is a reference for that field in the exception record, a ValidationError is raised based on the exception data.

```python
from typing import Any
from pydantic import BaseModel, Field, ValidationError, model_validator

def validate_model_field(model: type[BaseModel], field: str, value: Any) -> Any:
    if field not in model.model_fields:
        raise ValueError(f"'{field}' does not denote a field of model '{model}'.")

    try:
        model(**{field: value})
    except ValidationError as e:
        for error in e.errors():
            if field in error["loc"]:
                raise ValidationError.from_exception_data(model.__name__, [error])

    return value
```

I'm not exactly sure if this might be inefficient for more involved models, because apparently Pydantic runs the entire model validation and collects exceptions, so no fast failure. But unlike the accepted answer, this handles models with model validators.

I would really appreciate opinions regarding the proposed solution! :)

**Reply by jordemort** [reply]
I prefer your solution to mine, I've always thought mine was pretty ugly and hacky and I'm surprised other people like it

**Reply by shivam-khandelwal** [reply]
If anyone has found an official Pydantic solution, please do post here - seems like V2 went backwords on this!

**Comment by JoongWonSeo** [comment]
Hey guys, I also ran into this issue and have tried out all of these solutions in the thread, and ran into different limitations, esp. for the following difficult use case:
- Needs to validate arbitrarily complex fields with annotations, defaults, field validators, etc.
- Must raise when validation fails, must return validated value when it succeeds
- Needs to be fast and side-effect free (in my use case, some models trigger side-effects in `model_post_init()`)

```python
from typing import Annotated
from pydantic import AfterValidator, BaseModel, Field, field_validator, model_validator

type TrimmedStr = Annotated[str, AfterValidator(lambda v: v.strip())]

class Test(BaseModel):
    name: str = Field(min_length=2, max_length=10)
    """string of length 2-10"""

    email: str = Field(pattern=r"^[^@]+@[^@]+\.[^@]+$")
    """regex email validation"""

    trimmed: TrimmedStr
    """auto-strip whitespace from input"""

    percent: Annotated[float, Field(ge=0, le=1), AfterValidator(lambda v: v * 100)]
    """expect the input to be in 0-1, but we want to store it as 0-100"""

    reversed_str: TrimmedStr
    """auto-reversed by the below field validator"""

    @field_validator("reversed_str")
    @classmethod
    def reverse_str(cls, v: str) -> str:
        return v[::-1]

    @model_validator(mode="after")
    def email_contains_name(self):
        if self.name not in self.email:
            raise ValueError(f"email should contain name: {self.name}")
        return self

    def model_post_init(self, context):
        print("I am a side effect triggered by the model init")


Test(
    name="jay",
    email="jay@test.com",
    trimmed="   hello  ",
    percent=0.5,
    reversed_str="   hello  ",
)
```
Expected to produce an object like:
`Test(name='jay', email='jay@test.com', trimmed='hello', percent=50.0, reversed_str='olleh')`
While also printing `"I am a side effect triggered by the model init"` to console, due to the `model_post_init`.

1. @sydney-runkle's solution using TypeAdapter: Most elegant and side-effect free, but as pointed out, only does type coercion, loses most other validations (although some do work, actually).
```python
from pydantic import TypeAdapter

validator = TypeAdapter(Test.model_fields['reversed_str'].annotation)
print(validator.validate_python("  hello "))  # returns 'hello' instead of 'olleh'
%timeit validator.validate_python("  hello ")
```
Output:
```
hello
187 ns ± 2.27 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

2. @jordemort's solution works well, but **1)** requires creating an instance of the model, which triggers side-effects, and **2)** as pointed out by @lu-pl, it will fail when using `@model_validator`s.
However, it does return an instance with that single field assigned, so the following does work (when taking out the `@model_validator`):
```python
cache: dict[type[BaseModel], Any] = {}

def validate_model_field(model: type[BaseModel], field_name: str, value: Any) -> Any:
    if (m:=cache.get(model)) is None:
        m = model.model_construct()
        cache[model] = m

    return getattr(model.__pydantic_validator__.validate_assignment(m, field_name, value), field_name)


print(validate_model_field(Test, "reversed_str", "  hello "))
%timeit validate_model_field(Test, "reversed_str", "  hello ")
```
Output:
```
olleh
604 ns ± 7.71 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

3. @lu-pl 's solution works even with `@model_validator`s but unfortunately I couldn't find any way to return the validated value instead of returning the raw value, losing type coercion/conversion and any transformations from validations. Interestingly, since the `ValidationError` is thrown most of the cases, it won't trigger side-effects in my `model_post_init` unless the validated field is the only required field.

```python
def validate_model_field(model: type[BaseModel], field: str, value: Any) -> Any:
    if field not in model.model_fields:
        raise ValueError(f"'{field}' does not denote a field of model '{model}'.")

    try:
        model(**{field: value})
    except ValidationError as e:
        for error in e.errors():
            if field in error["loc"]:
                raise ValidationError.from_exception_data(model.__name__, [error])

    return value

print(validate_model_field(Test, "reversed_str", "  hello "))
%timeit validate_model_field(Test, "reversed_str", "  hello ")
```
Output:
```
  hello 
3.23 μs ± 38.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)`
```

I ended up using the low-level API and basically reconstructing the `CoreSchema` to include only the field's schema + any potential external definitions (used for validators apparently):

```python
from functools import cache
from pydantic_core import CoreSchema, SchemaValidator


def find_field_schema(model: type[BaseModel], field_name: str) -> CoreSchema:
    schema: CoreSchema = model.__pydantic_core_schema__.copy()
    # we shallow copied, be careful not to mutate the original schema!

    assert schema["type"] in ["definitions", "model"]

    # find the field schema
    field_schema = schema["schema"] # type: ignore
    while "fields" not in field_schema:
        field_schema = field_schema["schema"] # type: ignore
    
    field_schema = field_schema["fields"][field_name]["schema"] # type: ignore

    # if the original schema is a definition schema, replace the model schema with the field schema
    if schema["type"] == "definitions":
        schema["schema"] = field_schema
        return schema
    else:
        return field_schema

@cache
def validator(model: type[BaseModel], field_name: str) -> SchemaValidator:
    return SchemaValidator(find_field_schema(model, field_name))

def validate_model_field(model: type[BaseModel], field_name: str, value: Any) -> Any:
    return validator(model, field_name).validate_python(value)


print(validate_model_field(Test, "reversed_str", "  hello "))
%timeit validate_model_field(Test, "reversed_str", "  hello ")
```
Output:
```
olleh
201 ns ± 3.96 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
